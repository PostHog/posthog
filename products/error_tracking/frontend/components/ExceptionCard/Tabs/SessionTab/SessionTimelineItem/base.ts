import { IconComponent } from '@posthog/icons/dist/src/types/icon-types'

export interface PreviewRenderProps<T> {
    item: T
    selected: boolean
}

export interface DetailsRenderProps<T> extends PreviewRenderProps<T> {}

export type SessionTimelineEvent = {
    id: string
    type: 'event'
    timestamp: string
    payload: {
        event: string
        properties: Record<string, any>
    }
}

export type SessionTimelineItem = SessionTimelineEvent

export enum RendererGroup {
    ERROR_TRACKING = 'error-tracking',
    PRODUCT_ANALYTICS = 'product-analytics',
    WEB_ANALYTICS = 'web-analytics',
    SURVEYS = 'surveys',
    FEATURE_FLAGS = 'feature-flags',
    INTERNALS = 'internals',
}

export type SessionTimelineRenderer<T> = {
    predicate: (item: SessionTimelineItem) => boolean
    runtimeIcon: IconComponent<PreviewRenderProps<T>>
    group: RendererGroup
    renderPreview: React.FC<PreviewRenderProps<T>>
    renderDetails: React.FC<DetailsRenderProps<T>>
}
