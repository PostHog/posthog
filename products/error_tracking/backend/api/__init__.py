from .assignment_rules import ErrorTrackingAssignmentRuleViewSet
from .external_references import ErrorTrackingExternalReferenceViewSet
from .fingerprints import ErrorTrackingFingerprintViewSet
from .git_provider_file_link_resolver import GitProviderFileLinksViewSet
from .grouping_rules import ErrorTrackingGroupingRuleViewSet
from .issues import ErrorTrackingIssueViewSet
from .releases import ErrorTrackingReleaseViewSet
from .stack_frames import ErrorTrackingStackFrameViewSet
from .suppression_rules import ErrorTrackingSuppressionRuleViewSet
from .symbol_sets import ErrorTrackingSymbolSetViewSet

__all__ = [
    "ErrorTrackingExternalReferenceViewSet",
    "ErrorTrackingIssueViewSet",
    "ErrorTrackingStackFrameViewSet",
    "ErrorTrackingSymbolSetViewSet",
    "ErrorTrackingFingerprintViewSet",
    "ErrorTrackingGroupingRuleViewSet",
    "ErrorTrackingReleaseViewSet",
    "ErrorTrackingSuppressionRuleViewSet",
    "ErrorTrackingAssignmentRuleViewSet",
    "GitProviderFileLinksViewSet",
]
