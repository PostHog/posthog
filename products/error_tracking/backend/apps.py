from django.apps import AppConfig


class ErrorTrackingConfig(AppConfig):
    default_auto_field = "django.db.models.BigAutoField"
    name = "products.error_tracking.backend"
    label = "error_tracking"
