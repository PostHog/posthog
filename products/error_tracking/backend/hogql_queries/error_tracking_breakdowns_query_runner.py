import datetime
from zoneinfo import ZoneInfo

import structlog

from posthog.schema import (
    CachedErrorTrackingBreakdownsQueryResponse,
    ErrorTrackingBreakdownsQuery,
    ErrorTrackingBreakdownsQueryResponse,
)

from posthog.hogql import ast
from posthog.hogql.query import execute_hogql_query

from posthog.hogql_queries.query_runner import AnalyticsQueryRunner
from posthog.queries.trends.breakdown import BREAKDOWN_NULL_STRING_LABEL
from posthog.utils import relative_date_parse

logger = structlog.get_logger(__name__)


class ErrorTrackingBreakdownsQueryRunner(AnalyticsQueryRunner[ErrorTrackingBreakdownsQueryResponse]):
    query: ErrorTrackingBreakdownsQuery
    cached_response: CachedErrorTrackingBreakdownsQueryResponse

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.date_from = self.parse_relative_date_from(self.query.dateRange.date_from if self.query.dateRange else None)
        self.date_to = self.parse_relative_date_to(self.query.dateRange.date_to if self.query.dateRange else None)

    @classmethod
    def parse_relative_date_from(cls, date: str | None) -> datetime.datetime:
        if date == "all" or date is None:
            return datetime.datetime.now(tz=ZoneInfo("UTC")) - datetime.timedelta(days=7)

        return relative_date_parse(date, now=datetime.datetime.now(tz=ZoneInfo("UTC")), timezone_info=ZoneInfo("UTC"))

    @classmethod
    def parse_relative_date_to(cls, date: str | None) -> datetime.datetime:
        if not date:
            return datetime.datetime.now(tz=ZoneInfo("UTC"))
        if date == "all":
            raise ValueError("Invalid date range")

        return relative_date_parse(date, ZoneInfo("UTC"), increase=True)

    def to_query(self) -> ast.SelectQuery:
        array_elements: list[ast.Expr] = []
        for prop in self.query.breakdownProperties:
            tuple_elements = [
                ast.Constant(value=prop),
                ast.Call(
                    name="ifNull",
                    args=[
                        ast.Call(name="toString", args=[ast.Field(chain=["properties", prop])]),
                        ast.Constant(value=BREAKDOWN_NULL_STRING_LABEL),
                    ],
                ),
            ]
            array_elements.append(ast.Tuple(exprs=tuple_elements))

        tuples_select = ast.SelectQuery(
            select=[
                ast.Alias(
                    alias="breakdown_tuple",
                    expr=ast.Call(name="arrayJoin", args=[ast.Array(exprs=array_elements)]),
                )
            ],
            select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
            where=self.tuples_select_where(),
        )

        tuples_unpack_select = ast.SelectQuery(
            select=[
                ast.Alias(
                    alias="breakdown_property",
                    expr=ast.Call(
                        name="tupleElement", args=[ast.Field(chain=["breakdown_tuple"]), ast.Constant(value=1)]
                    ),
                ),
                ast.Alias(
                    alias="breakdown_value",
                    expr=ast.Call(
                        name="tupleElement", args=[ast.Field(chain=["breakdown_tuple"]), ast.Constant(value=2)]
                    ),
                ),
            ],
            select_from=ast.JoinExpr(table=tuples_select),
        )

        with_total_count_select = ast.SelectQuery(
            select=[
                ast.Field(chain=["breakdown_property"]),
                ast.Field(chain=["breakdown_value"]),
                ast.Alias(alias="count", expr=ast.Call(name="count", args=[])),
                ast.Alias(
                    alias="total_count",
                    expr=ast.WindowFunction(
                        name="sum",
                        args=[ast.Field(chain=["count"])],
                        over_expr=ast.WindowExpr(
                            partition_by=[ast.Field(chain=["breakdown_property"])],
                        ),
                    ),
                ),
            ],
            select_from=ast.JoinExpr(table=tuples_unpack_select),
            group_by=[ast.Field(chain=["breakdown_property"]), ast.Field(chain=["breakdown_value"])],
        )

        with_row_number_select = ast.SelectQuery(
            select=[
                ast.Field(chain=["breakdown_property"]),
                ast.Field(chain=["breakdown_value"]),
                ast.Field(chain=["count"]),
                ast.Field(chain=["total_count"]),
                ast.Alias(
                    alias="rn",
                    expr=ast.WindowFunction(
                        name="row_number",
                        args=[],
                        over_expr=ast.WindowExpr(
                            partition_by=[ast.Field(chain=["breakdown_property"])],
                            order_by=[ast.OrderExpr(expr=ast.Field(chain=["count"]), order="DESC")],
                        ),
                    ),
                ),
            ],
            select_from=ast.JoinExpr(table=with_total_count_select),
        )

        limit_value = self.query.limit if self.query.limit is not None else 3
        final_select = ast.SelectQuery(
            select=[
                ast.Field(chain=["breakdown_property"]),
                ast.Field(chain=["breakdown_value"]),
                ast.Field(chain=["count"]),
                ast.Field(chain=["total_count"]),
            ],
            select_from=ast.JoinExpr(table=with_row_number_select),
            where=ast.CompareOperation(
                left=ast.Field(chain=["rn"]), right=ast.Constant(value=limit_value), op=ast.CompareOperationOp.LtEq
            ),
            order_by=[
                ast.OrderExpr(expr=ast.Field(chain=["breakdown_property"]), order="ASC"),
                ast.OrderExpr(expr=ast.Field(chain=["count"]), order="DESC"),
            ],
        )

        return final_select

    def tuples_select_where(self) -> ast.Expr:
        conditions: list[ast.Expr] = []

        conditions.append(
            ast.CompareOperation(
                left=ast.Field(chain=["timestamp"]),
                right=ast.Constant(value=self.date_from),
                op=ast.CompareOperationOp.GtEq,
            )
        )
        conditions.append(
            ast.CompareOperation(
                left=ast.Field(chain=["timestamp"]),
                right=ast.Constant(value=self.date_to),
                op=ast.CompareOperationOp.LtEq,
            )
        )

        conditions.append(
            ast.CompareOperation(
                left=ast.Field(chain=["event"]), right=ast.Constant(value="$exception"), op=ast.CompareOperationOp.Eq
            )
        )

        conditions.append(
            ast.CompareOperation(
                left=ast.Field(chain=["properties", "$exception_issue_id"]),
                right=ast.Constant(value=self.query.issueId),
                op=ast.CompareOperationOp.Eq,
            )
        )

        if self.query.filterTestAccounts:
            conditions.append(
                ast.CompareOperation(
                    left=ast.Call(
                        name="ifNull",
                        args=[
                            ast.Field(chain=["properties", "$is_identified"]),
                            ast.Constant(value=False),
                        ],
                    ),
                    right=ast.Constant(value=True),
                    op=ast.CompareOperationOp.Eq,
                )
            )

        return ast.And(exprs=conditions)

    def _calculate(self):
        with self.timings.measure("error_tracking_breakdowns_hogql_execute"):
            query_result = execute_hogql_query(
                query=self.to_query(),
                team=self.team,
                query_type="ErrorTrackingBreakdownsQuery",
                timings=self.timings,
                modifiers=self.modifiers,
                limit_context=self.limit_context,
            )

        grouped_results: dict[str, dict] = {}
        for row in query_result.results:
            breakdown_property = str(row[0])
            breakdown_value = str(row[1])
            count = int(row[2])
            total_count = int(row[3])

            if breakdown_property not in grouped_results:
                grouped_results[breakdown_property] = {"values": [], "total_count": total_count}

            grouped_results[breakdown_property]["values"].append({"breakdown_value": breakdown_value, "count": count})

        return ErrorTrackingBreakdownsQueryResponse(
            results=grouped_results,
            timings=query_result.timings,
            hogql=query_result.hogql,
            modifiers=self.modifiers,
        )
