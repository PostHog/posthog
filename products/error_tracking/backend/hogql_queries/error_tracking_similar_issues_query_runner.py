from datetime import datetime, timedelta

from django.db.models.aggregates import Max

import structlog

from posthog.schema import (
    CachedErrorTrackingSimilarIssuesQueryResponse,
    EmbeddingModelName,
    ErrorTrackingQueryResponse,
    ErrorTrackingSimilarIssuesQuery,
    ErrorTrackingSimilarIssuesQueryResponse,
    HogQLQueryResponse,
    SimilarIssue,
)

from posthog.hogql import ast
from posthog.hogql.constants import LimitContext
from posthog.hogql.parser import parse_select
from posthog.hogql.query import execute_hogql_query

from posthog.hogql_queries.insights.paginators import HogQLHasMorePaginator
from posthog.hogql_queries.query_runner import AnalyticsQueryRunner

from products.error_tracking.backend.api.issues import ErrorTrackingIssuePreviewSerializer
from products.error_tracking.backend.models import ErrorTrackingIssue, ErrorTrackingIssueFingerprintV2

logger = structlog.get_logger(__name__)

DOCUMENT_TYPE = "fingerprint"


class ErrorTrackingSimilarIssuesQueryRunner(AnalyticsQueryRunner[ErrorTrackingQueryResponse]):
    query: ErrorTrackingSimilarIssuesQuery
    cached_response: CachedErrorTrackingSimilarIssuesQueryResponse
    paginator: HogQLHasMorePaginator

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.paginator = HogQLHasMorePaginator.from_limit_context(
            limit_context=LimitContext.QUERY,
            limit=self.query.limit if self.query.limit else None,
            offset=self.query.offset,
        )
        ## Validate query

    def to_query(self) -> ast.SelectQuery | ast.SelectSetQuery:
        matched_fingerprints = list(
            ErrorTrackingIssueFingerprintV2.objects.filter(
                team=self.team,
                issue_id=self.query.issueId,
            )
            .annotate(mversion=Max("version"))
            .values("fingerprint")
        )
        target_fingerprints = [fingerprint["fingerprint"] for fingerprint in matched_fingerprints]
        return parse_select(
            self.query_template,
            placeholders={
                "fingerprints": ast.Constant(value=target_fingerprints),
                "model_name": ast.Constant(value=self.model_name),
                "rendering": ast.Constant(value=self.rendering),
                "max_distance": ast.Constant(value=self.max_distance),
            },
        )

    def _calculate(self):
        with self.timings.measure("error_tracking_similar_issues_hogql_execute"):
            query_result = self.paginator.execute_hogql_query(
                query=self.to_query(),
                team=self.team,
                query_type="ErrorTrackingSimilarIssuesQuery",
                timings=self.timings,
                modifiers=self.modifiers,
                limit_context=self.limit_context,
            )
        similar_fingerprints = [row[0] for row in query_result.results]
        similar_issues = self.fetch_issues(similar_fingerprints).data
        first_event_data = self.fetch_first_event_data(similar_issues)
        return ErrorTrackingSimilarIssuesQueryResponse(
            results=[
                SimilarIssue(
                    id=issue.get("id"),
                    name=issue.get("name"),
                    description=issue.get("description"),
                    first_seen=issue.get("first_seen"),
                    status=issue.get("status"),
                    library=first_event_data.get(issue.get("id"), None),
                )
                for issue in similar_issues
            ],
            timings=query_result.timings,
            hogql=query_result.hogql,
            modifiers=self.modifiers,
            **self.paginator.response_params(),
        )

    def fetch_issues(self, similar_fingerprints: list[str]) -> ErrorTrackingIssuePreviewSerializer:
        issue_queryset = (
            ErrorTrackingIssue.objects.with_first_seen()
            .filter(team=self.team, fingerprints__fingerprint__in=similar_fingerprints)
            .distinct()
        )
        return ErrorTrackingIssuePreviewSerializer(issue_queryset, many=True)

    def fetch_first_event_data(self, similar_issues) -> dict[str, str]:
        time_points: list[datetime] = [datetime.fromisoformat(issue.get("first_seen")) for issue in similar_issues]
        time_window = timedelta(hours=1)
        time_conditions: list[ast.Expr] = [
            ast.And(
                exprs=[
                    ast.CompareOperation(
                        left=ast.Field(chain=["timestamp"]),
                        right=ast.Constant(value=time_point - time_window),
                        op=ast.CompareOperationOp.GtEq,
                    ),
                    ast.CompareOperation(
                        left=ast.Field(chain=["timestamp"]),
                        right=ast.Constant(value=time_point + time_window),
                        op=ast.CompareOperationOp.LtEq,
                    ),
                ]
            )
            for time_point in time_points
        ]
        global_time_conditions = ast.Or(exprs=time_conditions)
        query = ast.SelectQuery(
            select=[
                ast.Field(chain=["properties", "$exception_issue_id"]),
                ast.Call(name="Min", args=[ast.Field(chain=["properties", "$lib"])]),
            ],
            select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
            where=ast.And(
                exprs=[
                    global_time_conditions,
                    ast.CompareOperation(
                        left=ast.Field(chain=["event"]),
                        right=ast.Constant(value="$exception"),
                        op=ast.CompareOperationOp.Eq,
                    ),
                ],
            ),
            group_by=[ast.Field(chain=["properties", "$exception_issue_id"])],
        )
        results: HogQLQueryResponse = execute_hogql_query(query, team=self.team)
        result_dict = {}
        for row in results.results:
            result_dict[row[0]] = row[1]
        return result_dict

    @property
    def query_template(self):
        return """
        SELECT
            b.document_id as fingerprint,
            MIN(cosineDistance(a.embedding, b.embedding)) as distance
        FROM
        (
            SELECT document_id, document_type, rendering, model_name, embedding
            FROM document_embeddings
            WHERE document_type = 'fingerprint'
            AND rendering = {rendering}
            AND model_name = {model_name}
            AND document_id IN {fingerprints}
            AND product = 'error_tracking'
        ) as a
        CROSS JOIN (
            SELECT document_id, document_type, rendering, model_name, embedding
            FROM document_embeddings
            WHERE document_type = 'fingerprint'
            AND rendering = {rendering}
            AND model_name = {model_name}
            AND document_id NOT IN {fingerprints}
            AND product = 'error_tracking'
        ) as b
        GROUP BY fingerprint
        HAVING distance <= {max_distance}
        ORDER BY distance ASC
        """

    @property
    def model_name(self):
        return self.query.modelName or str(EmbeddingModelName.TEXT_EMBEDDING_3_LARGE_3072)

    @property
    def rendering(self):
        return self.query.rendering or "type_message_and_stack"

    @property
    def max_distance(self):
        return self.query.maxDistance or 0.1
