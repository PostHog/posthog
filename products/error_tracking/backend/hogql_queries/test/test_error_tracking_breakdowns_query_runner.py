from freezegun import freeze_time
from posthog.test.base import (
    APIBaseTest,
    ClickhouseTestMixin,
    _create_event,
    flush_persons_and_events,
    snapshot_clickhouse_queries,
)

from posthog.schema import DateRange, ErrorTrackingBreakdownsQuery

from products.error_tracking.backend.hogql_queries.error_tracking_breakdowns_query_runner import (
    ErrorTrackingBreakdownsQueryRunner,
)
from products.error_tracking.backend.models import ErrorTrackingIssue, ErrorTrackingIssueFingerprintV2


class TestErrorTrackingBreakdownsQueryRunner(ClickhouseTestMixin, APIBaseTest):
    issue_id = "01936e7f-d7ff-7314-b2d4-7627981e34f0"
    fingerprint = "test_fingerprint"

    def create_issue(self, issue_id, fingerprint):
        issue = ErrorTrackingIssue.objects.create(id=issue_id, team=self.team)
        ErrorTrackingIssueFingerprintV2.objects.create(team=self.team, issue=issue, fingerprint=fingerprint)
        return issue

    def create_exception_event(self, distinct_id, properties):
        base_properties = {
            "$exception_issue_id": self.issue_id,
            "$exception_fingerprint": self.fingerprint,
        }
        _create_event(
            distinct_id=distinct_id,
            event="$exception",
            team=self.team,
            properties={**base_properties, **properties},
        )

    @freeze_time("2024-01-10T12:00:00Z")
    @snapshot_clickhouse_queries
    def test_breakdown_with_limit(self):
        self.create_issue(self.issue_id, self.fingerprint)

        browsers = [("A", 10), ("B", 8), ("C", 6), ("D", 4), ("E", 2), ("F", 1)]
        for browser, count in browsers:
            for _ in range(count):
                self.create_exception_event(f"user_{browser}", {"$browser": browser})

        flush_persons_and_events()

        runner = ErrorTrackingBreakdownsQueryRunner(
            team=self.team,
            query=ErrorTrackingBreakdownsQuery(
                kind="ErrorTrackingBreakdownsQuery",
                issueId=self.issue_id,
                breakdownProperties=["$browser"],
                dateRange=DateRange(date_from="-7d"),
                limit=3,
            ),
        )

        response = runner.calculate()

        browser_data = response.results["$browser"]

        assert len(browser_data.values) == 3
        assert browser_data.total_count == 31

        assert browser_data.values[0].breakdown_value == "A"
        assert browser_data.values[0].count == 10
        assert browser_data.values[1].breakdown_value == "B"
        assert browser_data.values[1].count == 8
        assert browser_data.values[2].breakdown_value == "C"
        assert browser_data.values[2].count == 6

    @freeze_time("2024-01-10T12:00:00Z")
    @snapshot_clickhouse_queries
    def test_multiple_breakdown_properties(self):
        self.create_issue(self.issue_id, self.fingerprint)

        for _ in range(5):
            self.create_exception_event("user_1", {"$browser": "Chrome", "$os": "Windows"})
        for _ in range(3):
            self.create_exception_event("user_2", {"$browser": "Firefox", "$os": "macOS"})
        for _ in range(2):
            self.create_exception_event("user_3", {"$browser": "Safari", "$os": "macOS"})

        flush_persons_and_events()

        runner = ErrorTrackingBreakdownsQueryRunner(
            team=self.team,
            query=ErrorTrackingBreakdownsQuery(
                kind="ErrorTrackingBreakdownsQuery",
                issueId=self.issue_id,
                breakdownProperties=["$browser", "$os"],
                dateRange=DateRange(date_from="-7d"),
                limit=3,
            ),
        )

        response = runner.calculate()

        assert "$browser" in response.results
        assert "$os" in response.results

        browser_data = response.results["$browser"]
        assert len(browser_data.values) == 3
        assert browser_data.total_count == 10
        assert browser_data.values[0].breakdown_value == "Chrome"
        assert browser_data.values[0].count == 5

        os_data = response.results["$os"]
        assert len(os_data.values) == 2
        assert os_data.total_count == 10
        assert os_data.values[0].breakdown_value == "macOS"
        assert os_data.values[0].count == 5
        assert os_data.values[1].breakdown_value == "Windows"
        assert os_data.values[1].count == 5

    @freeze_time("2024-01-10T12:00:00Z")
    @snapshot_clickhouse_queries
    def test_breakdown_with_null_values(self):
        self.create_issue(self.issue_id, self.fingerprint)

        for _ in range(5):
            self.create_exception_event("user_1", {"$browser": "Chrome"})
        for _ in range(3):
            self.create_exception_event("user_2", {})

        flush_persons_and_events()

        runner = ErrorTrackingBreakdownsQueryRunner(
            team=self.team,
            query=ErrorTrackingBreakdownsQuery(
                kind="ErrorTrackingBreakdownsQuery",
                issueId=self.issue_id,
                breakdownProperties=["$browser"],
                dateRange=DateRange(date_from="-7d"),
                limit=3,
            ),
        )

        response = runner.calculate()

        browser_data = response.results["$browser"]
        assert len(browser_data.values) == 2
        assert browser_data.total_count == 8

        assert browser_data.values[0].breakdown_value == "Chrome"
        assert browser_data.values[0].count == 5

        assert browser_data.values[1].breakdown_value == "$$_posthog_breakdown_null_$$"
        assert browser_data.values[1].count == 3

    @freeze_time("2024-01-10T12:00:00Z")
    @snapshot_clickhouse_queries
    def test_breakdown_respects_date_range(self):
        self.create_issue(self.issue_id, self.fingerprint)

        # outside range
        with freeze_time("2024-01-01T12:00:00Z"):
            for _ in range(10):
                self.create_exception_event("user_old", {"$browser": "OldBrowser"})

        # within range
        for _ in range(5):
            self.create_exception_event("user_1", {"$browser": "Chrome"})

        flush_persons_and_events()

        runner = ErrorTrackingBreakdownsQueryRunner(
            team=self.team,
            query=ErrorTrackingBreakdownsQuery(
                kind="ErrorTrackingBreakdownsQuery",
                issueId=self.issue_id,
                breakdownProperties=["$browser"],
                dateRange=DateRange(date_from="-7d"),
                limit=3,
            ),
        )

        response = runner.calculate()

        browser_data = response.results["$browser"]

        assert len(browser_data.values) == 1
        assert browser_data.total_count == 5
        assert browser_data.values[0].breakdown_value == "Chrome"
        assert browser_data.values[0].count == 5
