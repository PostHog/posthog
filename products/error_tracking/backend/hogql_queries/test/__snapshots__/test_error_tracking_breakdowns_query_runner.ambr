# serializer version: 1
# name: TestErrorTrackingBreakdownsQueryRunner.test_breakdown_respects_date_range
  '''
  SELECT breakdown_property AS breakdown_property,
         breakdown_value AS breakdown_value,
         count AS count,
                  total_count AS total_count
  FROM
    (SELECT breakdown_property AS breakdown_property,
            breakdown_value AS breakdown_value,
            count AS count,
                     total_count AS total_count,
                     row_number() OVER (PARTITION BY breakdown_property
                                        ORDER BY count DESC) AS rn
     FROM
       (SELECT breakdown_property AS breakdown_property,
               breakdown_value AS breakdown_value,
               count() AS count,
               sum()(count) OVER (PARTITION BY breakdown_property) AS total_count
        FROM
          (SELECT tupleElement(breakdown_tuple, 1) AS breakdown_property,
                  tupleElement(breakdown_tuple, 2) AS breakdown_value
           FROM
             (SELECT arrayJoin([tuple('$browser', ifNull(toString(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, '$browser'), ''), 'null'), '^"|"$', '')), '$$_posthog_breakdown_null_$$'))]) AS breakdown_tuple
              FROM events
              WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), toDateTime64('2024-01-03 12:00:00.000000', 6, 'UTC')), lessOrEquals(toTimeZone(events.timestamp, 'UTC'), toDateTime64('today', 6, 'UTC')), equals(events.event, '$exception'), ifNull(equals(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, '$exception_issue_id'), ''), 'null'), '^"|"$', ''), '01936e7f-d7ff-7314-b2d4-7627981e34f0'), 0))))
        GROUP BY breakdown_property,
                 breakdown_value))
  WHERE ifNull(lessOrEquals(rn, 3), 0)
  ORDER BY breakdown_property ASC,
           count DESC
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295,
                     allow_experimental_join_condition=1
  '''
# ---
# name: TestErrorTrackingBreakdownsQueryRunner.test_breakdown_with_limit
  '''
  SELECT breakdown_property AS breakdown_property,
         breakdown_value AS breakdown_value,
         count AS count,
                  total_count AS total_count
  FROM
    (SELECT breakdown_property AS breakdown_property,
            breakdown_value AS breakdown_value,
            count AS count,
                     total_count AS total_count,
                     row_number() OVER (PARTITION BY breakdown_property
                                        ORDER BY count DESC) AS rn
     FROM
       (SELECT breakdown_property AS breakdown_property,
               breakdown_value AS breakdown_value,
               count() AS count,
               sum()(count) OVER (PARTITION BY breakdown_property) AS total_count
        FROM
          (SELECT tupleElement(breakdown_tuple, 1) AS breakdown_property,
                  tupleElement(breakdown_tuple, 2) AS breakdown_value
           FROM
             (SELECT arrayJoin([tuple('$browser', ifNull(toString(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, '$browser'), ''), 'null'), '^"|"$', '')), '$$_posthog_breakdown_null_$$'))]) AS breakdown_tuple
              FROM events
              WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), toDateTime64('2024-01-03 12:00:00.000000', 6, 'UTC')), lessOrEquals(toTimeZone(events.timestamp, 'UTC'), toDateTime64('today', 6, 'UTC')), equals(events.event, '$exception'), ifNull(equals(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, '$exception_issue_id'), ''), 'null'), '^"|"$', ''), '01936e7f-d7ff-7314-b2d4-7627981e34f0'), 0))))
        GROUP BY breakdown_property,
                 breakdown_value))
  WHERE ifNull(lessOrEquals(rn, 3), 0)
  ORDER BY breakdown_property ASC,
           count DESC
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295,
                     allow_experimental_join_condition=1
  '''
# ---
# name: TestErrorTrackingBreakdownsQueryRunner.test_breakdown_with_null_values
  '''
  SELECT breakdown_property AS breakdown_property,
         breakdown_value AS breakdown_value,
         count AS count,
                  total_count AS total_count
  FROM
    (SELECT breakdown_property AS breakdown_property,
            breakdown_value AS breakdown_value,
            count AS count,
                     total_count AS total_count,
                     row_number() OVER (PARTITION BY breakdown_property
                                        ORDER BY count DESC) AS rn
     FROM
       (SELECT breakdown_property AS breakdown_property,
               breakdown_value AS breakdown_value,
               count() AS count,
               sum()(count) OVER (PARTITION BY breakdown_property) AS total_count
        FROM
          (SELECT tupleElement(breakdown_tuple, 1) AS breakdown_property,
                  tupleElement(breakdown_tuple, 2) AS breakdown_value
           FROM
             (SELECT arrayJoin([tuple('$browser', ifNull(toString(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, '$browser'), ''), 'null'), '^"|"$', '')), '$$_posthog_breakdown_null_$$'))]) AS breakdown_tuple
              FROM events
              WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), toDateTime64('2024-01-03 12:00:00.000000', 6, 'UTC')), lessOrEquals(toTimeZone(events.timestamp, 'UTC'), toDateTime64('today', 6, 'UTC')), equals(events.event, '$exception'), ifNull(equals(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, '$exception_issue_id'), ''), 'null'), '^"|"$', ''), '01936e7f-d7ff-7314-b2d4-7627981e34f0'), 0))))
        GROUP BY breakdown_property,
                 breakdown_value))
  WHERE ifNull(lessOrEquals(rn, 3), 0)
  ORDER BY breakdown_property ASC,
           count DESC
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295,
                     allow_experimental_join_condition=1
  '''
# ---
# name: TestErrorTrackingBreakdownsQueryRunner.test_multiple_breakdown_properties
  '''
  SELECT breakdown_property AS breakdown_property,
         breakdown_value AS breakdown_value,
         count AS count,
                  total_count AS total_count
  FROM
    (SELECT breakdown_property AS breakdown_property,
            breakdown_value AS breakdown_value,
            count AS count,
                     total_count AS total_count,
                     row_number() OVER (PARTITION BY breakdown_property
                                        ORDER BY count DESC) AS rn
     FROM
       (SELECT breakdown_property AS breakdown_property,
               breakdown_value AS breakdown_value,
               count() AS count,
               sum()(count) OVER (PARTITION BY breakdown_property) AS total_count
        FROM
          (SELECT tupleElement(breakdown_tuple, 1) AS breakdown_property,
                  tupleElement(breakdown_tuple, 2) AS breakdown_value
           FROM
             (SELECT arrayJoin([tuple('$browser', ifNull(toString(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, '$browser'), ''), 'null'), '^"|"$', '')), '$$_posthog_breakdown_null_$$')), tuple('$os', ifNull(toString(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, '$os'), ''), 'null'), '^"|"$', '')), '$$_posthog_breakdown_null_$$'))]) AS breakdown_tuple
              FROM events
              WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), toDateTime64('2024-01-03 12:00:00.000000', 6, 'UTC')), lessOrEquals(toTimeZone(events.timestamp, 'UTC'), toDateTime64('today', 6, 'UTC')), equals(events.event, '$exception'), ifNull(equals(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, '$exception_issue_id'), ''), 'null'), '^"|"$', ''), '01936e7f-d7ff-7314-b2d4-7627981e34f0'), 0))))
        GROUP BY breakdown_property,
                 breakdown_value))
  WHERE ifNull(lessOrEquals(rn, 3), 0)
  ORDER BY breakdown_property ASC,
           count DESC
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295,
                     allow_experimental_join_condition=1
  '''
# ---
